src/component.ts

```diff
@@ -1,51 +1,155 @@
 export interface Component {
   id: string;
-  render(contextualStyle: string | null): HTMLElement;
+  /// Returns true if the views associated with this component should be updated based on the state update.
+  needsRender(stateUpdate: StateUpdate): boolean;
+  /// Customize the HTMLElement & Style associated with this Component
+  /// Returns child components, if any
+  render(
+    store: Store,
+    element: HTMLElement,
+    style: Style,
+  ): Component[];
 }
 
-export function updateStyle(
-  element: HTMLElement,
-  id: string,
-  contextualStyle: string | null,
-  style: string,
-) {
-  const styleId = `${id}-style`;
-  let localStyle = document.getElementById(styleId) as HTMLStyleElement;
-  if (!localStyle) {
-    localStyle = document.createElement("style");
-    localStyle.id = styleId;
-  }
-  localStyle.textContent = `
-  ${id} {
-    ${contextualStyle === null ? "" : contextualStyle}
-    ${style}
-  }
-  `;
-  element.append(localStyle);
+export class StateUpdate {
+  previousState: AppState;
+  currentState: AppState;
+  constructor(previousState: AppState, currentState: AppState) {
+    this.previousState = previousState;
+    this.currentState = currentState;
+  }
+}
+
+export class Style {
+  element: HTMLStyleElement;
+  id: string;
+  constructor(id: string, element: HTMLStyleElement) {
+    this.id = id;
+    this.element = element;
+  }
+  push(styling: StyleItem[]) {
+    let css = "";
+    this.element.textContent = "";
+    styling.forEach((styling) => {
+      css = css.concat(
+        `
+      ${styling.selector} {
+        ${styling.styling}
+      }
+      `,
+      );
+    });
+    this.element.textContent = css;
+  }
+}
+
+export class StyleItem {
+  selector: string;
+  styling: string;
+  constructor(selector: string, styling: string) {
+    this.selector = selector;
+    this.styling = styling;
+  }
 }
 
-export function updateElement(
-  component: Component,
-  children: HTMLElement[],
-  innerHTML: string | null,
-): HTMLElement {
-  const id = `${component.id}`;
-  let element = document.getElementById(id);
-  if (element) {
-    while (element.firstChild) {
-      element.removeChild(element.firstChild);
+export class AppState {
+  selectingFile = true;
+  file: File | null = null;
+  clone(): AppState {
+    const clone = new AppState();
+    clone.selectingFile = this.selectingFile;
+    clone.file = this.file;
+    return clone;
+  }
+}
+
+export class Store {
+  component: Component;
+  previousState: AppState | null = null;
+  state: AppState;
+  lastRenderedComponents: Set<string> = new Set();
+
+  constructor(state: AppState, component: Component) {
+    this.component = component;
+    this.state = state;
+  }
+
+  update() {
+    const currentRenderedComponents = new Set<string>();
+    this.updateComponent(this.component, currentRenderedComponents, null);
+
+    const unrenderedComponents = [...this.lastRenderedComponents].filter(
```

Comment:
Gotta give you all some history on this one, but for posterity here, @Alex6683-bot asked about the spread operator (...). In JavaScript, the spread operator is used with collections to put the elements of the collection into a new collection. On this line Cy's using square bracket short hand to create a new array, and then spreading the `lastRederedComponents` into that array to use the array instance method filter.